src/debugger/android/androidPlatform.ts

```diff
@@ -7,25 +7,69 @@ import {IRunOptions} from "../../common/launchArgs";
 import {CommandExecutor} from "../../common/commandExecutor";
 import {Package} from "../../common/node/package";
 import {PackageNameResolver} from "../../common/android/packageNameResolver";
+import {DeviceResolver, IDevice} from "../../common/android/deviceResolver";
 
 /**
  * Android specific platform implementation for debugging RN applications.
  */
 export class AndroidPlatform implements IAppPlatform {
 
-    public runApp(runOptions: IRunOptions): Q.Promise<void> {
-        return new CommandExecutor(runOptions.projectRoot).spawnAndWaitReactCommand("run-android");
-    }
+    private debugTarget: string;
+    private packageName: string;
 
-    public enableJSDebuggingMode(runOptions: IRunOptions): Q.Promise<void> {
+    public runApp(runOptions: IRunOptions): Q.Promise<void> {
         let pkg = new Package(runOptions.projectRoot);
-
-        return pkg.name()
+        return new CommandExecutor(runOptions.projectRoot).spawnAndWaitReactCommand("run-android")
+            .then(() => pkg.name())
             .then(appName => new PackageNameResolver(appName).resolvePackageName(runOptions.projectRoot))
             .then(packageName => {
-                let enableDebugCommand = `adb shell am broadcast -a "${packageName.toLowerCase()}.RELOAD_APP_ACTION" --ez jsproxy true`;
+                this.packageName = packageName;
                 let cexec = new CommandExecutor(runOptions.projectRoot);
-                return cexec.execute(enableDebugCommand);
+                let deviceResolver = new DeviceResolver();
+                return deviceResolver.getConnectedDevices()
+                    .then((devices: IDevice[]) => {
+                        if (devices.length > 1) {
+                            /* more than one device or emulator */
+                            this.debugTarget = this.getTargetEmulator(runOptions, devices);
+                            if (this.debugTarget) {
+                                /* Launching is needed only if we have more than one device active */
+                                let launchAppCommand = `adb -s ${this.debugTarget} shell am start -n ${packageName.toLowerCase()}/.MainActivity`;
```

Comment:
If you use this.packageName here, you can extract the .then starting on line 30, and put it at the same level of the .then of line 25 removing one indentation level.